### PR TITLE
[AI Chat] Add AI Chat redux

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds `aichatRedux.ts` to store values in the redux store for the AI Chat lab. Thus, the `ChatWorkspace` component is updated to use redux instead of local state.

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-14)

## Testing story
I tested locally to confirm that there were no changes after this updates in /allthethings AI Chat levels.

## Follow-up work
I added a [commit](https://github.com/code-dot-org/code-dot-org/pull/53404/commits/e861a7deba2dcd7bb2c4219726a0d44745a741c7) (which then I reverted) that added a `dispatch` to `clearChatHistory` when the 'Continue' button was clicked. This change did indeed clear chat history when a user clicked on the 'Continue' button to go to the next level. And I also check to see that Music lab levels in /allthethings  behaved as usual. 
Since AI Chat future is up in the air, I did not include this change in this PR but just making a note of it.

Other follow-up items based on review feedback:
- Block the user from submitting multiple messages in a row while waiting from a response
- Use a thunk in place of the onSubmit function
- Rethink the UI of new user message / refactor storing of user message while waiting for a chat completion response
- Add levelbuilder property `initialAssistantChatMessage` (not required and if left blank, use default message) [slack discussion](https://codedotorg.slack.com/archives/C052GPUCRNZ/p1692805333308489)

Other items that may not be addressed unless project is restarted:
- Displaying of `bot_description` on hover
- Error handling if user does not have correct permission (and error handling in general)
- Displaying an indicator or auto-scrolling when new messages are added to chat workspace [jira](https://codedotorg.atlassian.net/browse/LABS-16)
- Testing (unit tests, storybook)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (bencodeorg) [inline apps/src/aichat/views/ChatWorkspace.tsx:47]: `newUserMessage` isn't used anywhere currently, yeah? Can you elaborate on how you're envisioning using it?

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (bencodeorg) [inline apps/src/aichat/views/ChatWorkspace.tsx]: Done?

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (bencodeorg) [inline apps/src/aichat/views/ChatWorkspace.tsx:60]: Nice, I'll hook into this for the loading animation :)

**Review by bencodeorg — COMMENTED** [review]
Couple little comments but generally looks good to me!

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx:47]: @sanchitmalhotra126 and I discussed a while back displaying the message as it's being sent to backend in the workspace but with some kind of UI indicator that the response is not yet received and the user input message is not confirmed - maybe in a different color and/or with 'waiting...' appended to it.
Were you planning to add your 'waiting for response' icon in the user message editor input box?

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx]: Oh yeah - I'll remove comment -thanks!

File: apps/src/aichat/constants.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/constants.ts:9]: Probably a question for product/curriculum - do we know if this initial message is meant to be hardcoded, or set by levelbuilders? Or generated by open AI?

File: apps/src/aichat/redux/aichatRedux.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/redux/aichatRedux.ts:22]: Could be a separate change - but I think we probably also want to block the user from submitting multiple messages in a row while waiting from a response? Might need to check with product/curriculum, but if so, we can probably use the isWaitingForChatResponse state to prevent the user from submitting another message.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (sanchitmalhotra126) [reply apps/src/aichat/views/ChatWorkspace.tsx:47]: Yep! As an alternate idea - we could also consider adding the new user message to the list of `chatMessages` directly, but with a specific status like `PENDING` or `UNKNOWN` to indicate that we're waiting on a response? Then when we get the response back, we can flip the status to ok/inappropriate/etc. That might make rendering the list of ChatMessages a bit simpler - but this approach works too! With this, we specifically know which message we're waiting on a response for, which is good.

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks good overall! I'm still thinking about if it would make sense to use a thunk in place of the onSubmit function now that we have this redux store, but that's something we can explore/pair on later.

File: apps/src/aichat/views/ChatWorkspace.tsx
Comment (bencodeorg) [reply apps/src/aichat/views/ChatWorkspace.tsx:47]: Cool that makes sense! I was adding the loading gif to the main chat, like this:

image

**Review by bencodeorg — APPROVED** [review]

File: apps/src/aichat/constants.ts
Comment (fisher-alice) [reply apps/src/aichat/constants.ts:9]: Thanks! Just checked with prod/curric in [this Slack thread](https://codedotorg.slack.com/archives/C052GPUCRNZ/p1692805333308489) and will implement their suggestions/preferences.

File: apps/src/aichat/redux/aichatRedux.ts
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts:22]: Great idea! I'll implement in a follow-up. Thanks!

**Comment by fisher-alice** [comment]
> Looks good overall! I'm still thinking about if it would make sense to use a thunk in place of the onSubmit function now that we have this redux store, but that's something we can explore/pair on later.

Thanks Sanchit! I think using a thunk makes a lot of sense and definitely plan to implement in a follow-up. I would like to merge this sooner than later so I can unblock @bencodeorg 's work on implementing the 'AI is typing' UI.

**Pull request merged**
